File: today/remove-duplicates-from-sorted-list.ts
/**
 * https://leetcode.com/problems/remove-duplicates-from-sorted-list/description/
 */
export { deleteDuplicates, toList, toArray };

function deleteDuplicates(head: ListNode | null): ListNode | null {
  return null;
}

function toList(elements: number[]): ListNode | null {
  if (elements.length === 0) return null;
  let pointer: ListNode | null = null;
  for (let i = elements.length - 1; i >= 0; i--) {
    const node = new ListNode(elements[i]);
    node.next = pointer;
    pointer = node;
  }
  return pointer;
}

function toArray(list: ListNode | null): number[] {
  if (!list) return [];
  const result: number[] = [];
  while (list) {
    result.push(list.val);
    list = list.next;
  }
  return result;
}

class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

File: returned/python/today6/remove_duplicates_from_sorted_list.py
# https://leetcode.com/problems/remove-duplicates-from-sorted-list/description/
from typing import Optional, List

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def to_nodes(elements: List[int]) -> Optional[ListNode]:
    if not elements:
        return None
    prev = None
    for e in reversed(elements):
        node = ListNode(e)
        node.next = prev
        prev = node
    return prev

def to_array(head: Optional[ListNode]) -> List[int]:
    if not head:
        return []
    result = []
    while head:
        result.append(head.val)
        head = head.next
    return result

class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(None, head)
        p = dummy

        while p:
            if p.next and p.val == p.next.val:
                dup = p.next
                while dup.next and dup.val == dup.next.val:
                    dup = dup.next
                p.next = dup.next
            p = p.next

        return dummy.next